File: C++/225 - Implement Stack using Queues.cpp
/*
 225. Implement Stack using Queues (Easy)

 Notes:
     Store values "upside down" in queue to simulate the stack.
     Similar to Lc.232 Implement Queue using Stacks
*/

class MyStack {
private:
    queue<int> q, helper;
public:
    /** Initialize your data structure here. */
    MyStack() {
        
    }
    
    /** Push element x onto stack. */
    void push(int x) {
        while (!q.empty())
        {
            helper.push(q.front());
            q.pop();
        }
        q.push(x);
        while (!helper.empty())
        {
            q.push(helper.front());
            helper.pop();
        }
    }
    
    /** Removes the element on top of the stack and returns that element. */
    int pop() {
        int val = top();
        q.pop();
        return val;
    }
    
    /** Get the top element. */
    int top() {
        return q.front();
    }
    
    /** Returns whether the stack is empty. */
    bool empty() {
        return q.empty();
    }
};

File: JAVA/225 - Implement Stack using Queues.java
/*
 225. Implement Stack using Queues (E)

 
 Approach 1 : one queue
 Time: O(n)
 Space: O(n)

 Approach 2 : two queues
 Two queues A & B, A as stack style queue, B as the helper to make A stack style.
 Time: O(2n)
 Space: O(2n)

*/

/* Approach 1 */
class MyStack {
    Queue<Integer> q;

    /** Initialize your data structure here. */
    public MyStack() {
        q = new LinkedList<Integer>();
    }
    
    /** Push element x onto stack. */
    public void push(int x) {
        int n = q.size();
        q.add(x);
        for (int i = 0; i < n; i++) q.add(q.poll());
    }
    
    /** Removes the element on top of the stack and returns that element. */
    public int pop() {
        return q.poll();
    }
    
    /** Get the top element. */
    public int top() {
        return q.peek();
    }
    
    /** Returns whether the stack is empty. */
    public boolean empty() {
        return q.isEmpty();
    }
}


/* Approach 2 */
class MyStack {
    Queue<Integer> q;
    Queue<Integer> helper;

    /** Initialize your data structure here. */
    public MyStack() {
        q = new LinkedList<Integer>();
        helper = new LinkedList<Integer>();
    }
    
    /** Push element x onto stack. */
    public void push(int x) {
        helper.add(x);
        while (!q.isEmpty()) helper.add(q.poll());
        while (!helper.isEmpty()) q.add(helper.poll());
    }
    
    /** Removes the element on top of the stack and returns that element. */
    public int pop() {
        return q.poll();
    }
    
    /** Get the top element. */
    public int top() {
        return q.peek();
    }
    
    /** Returns whether the stack is empty. */
    public boolean empty() {
        return q.isEmpty();
    }
}